lang/funcs/core/os/exec_func.go

```diff
@@ -0,0 +1,164 @@
+// after the other, the downstream resources might not run for every line unless
+// the "Meta:realize" metaparam is set to true.
+type ExecFunc struct {
+	init *interfaces.Init
+
+	// Used to shutdown the node, killing the current process if needed.
+	// We only run one process at a time.
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// ArgGen returns the Nth arg name for this function.
+func (obj *ExecFunc) ArgGen(index int) (string, error) {
+	seq := []string{"shell_command"}
+	if l := len(seq); index >= l {
+		return "", fmt.Errorf("index %d exceeds arg length of %d", index, l)
+	}
+	return seq[index], nil
+}
+
+// Validate makes sure we've built our struct properly. It is usually unused for
+// normal functions that users can use directly.
+func (obj *ExecFunc) Validate() error {
+	return nil
+}
+
+// Info returns some static info about itself.
+func (obj *ExecFunc) Info() *interfaces.Info {
+	return &interfaces.Info{
+		Pure: false, // definitely false
+		Memo: false,
+		Sig:  types.NewType("func(shell_command str) str"),
+		Err:  obj.Validate(),
+	}
+}
+
+// Init runs some startup code for this function.
+func (obj *ExecFunc) Init(init *interfaces.Init) error {
+	obj.init = init
+	obj.ctx, obj.cancel = context.WithCancel(context.Background())
+	return nil
+}
+
+// Stream returns the changing values that this func has over time.
+func (obj *ExecFunc) Stream() error {
+	var wg sync.WaitGroup
+	defer wg.Wait()
+
+	// Close the output chan to signal that no more values are coming.
+	defer close(obj.init.Output)
+
+	// Kill the current process if it's still running.
+	defer obj.cancel()
+
+	for {
+		select {
+		case input, more := <-obj.init.Input:
+			if !more {
+				return nil
+			}
+			shellCommand := input.Struct()["shell_command"].Str()
+
+			// Run the command, connecting it to ctx so we can kill
+			// it if needed, and to two Readers so we can read its
+			// stdout and stderr.
+			cmd := exec.CommandContext(obj.ctx, "sh", "-c", shellCommand)
+			stdoutReader, err := cmd.StdoutPipe()
+			stderrReader, err := cmd.StderrPipe()
+			if err != nil {
+				return err
+			}
+			err = cmd.Start()
+			if err != nil {
+				return err
+			}
+
+			// Sends the command's stdout and stderr to the
+			// Readers. Terminates when the process exits, on its
+			// own or or due to obj.cancel().
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+
+				cmd.Wait()
+			}()
+
+			// Log the lines from stderr, to help the user debug.
+			// Terminates when the process exits, on its own or or
+			// due to obj.cancel().
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+
+				stderrScanner := bufio.NewScanner(stderrReader)
+				for stderrScanner.Scan() {
+					obj.init.Logf(
+						"ExecFunc: \"%v\": stderr: %v\n",
+						shellCommand,
+						stderrScanner.Text(),
+					)
+				}
+			}()
+
+			// Emit one value downstream for each line from stdout.
+			stdoutScanner := bufio.NewScanner(stdoutReader)
+			for stdoutScanner.Scan() {
```

Comment:
I'm assuming `stdoutScanner` and `stderrScanner` both close when the command exits right? Thus breaking us out of this infinite loop. If you've just once added a printf in there saying "this loop exited" or something, then that's perfect for me, and I would recommend leaving a short comment mentioning that the loop closes when the command exits, so future readers know it gets cleaned up early.

Reply:
I do know that the loop exists, not because of a printf but because this loop needs to terminate in order for the loop which encloses it to run its next iteration, which is needed in order to handle the next command in the input stream, and `os.exec(fmt.printf("echo %d %% 10 | bc", datetime.now()))` works as expected.

That being said, I tested this loop and the stderr one immediately above just to be sure, and they both terminate. I then added a comment to this loop.